src/lib/mina_lib/mina_lib.ml

```diff
@@ -1495,6 +1495,126 @@ let start_filtered_log ~commit_id
       () ;
     Ok () )
 
+let fetch_completed_snarks (module Context : CONTEXT) snark_pool network
+    top_block get_most_recent_valid_block =
+  let open Context in
+  let open Network_peer in
+  let%bind all_peers = Mina_networking.peers network in
+  Deferred.List.iter
+    ~f:(fun peer ->
+      [%log debug] "PEER IS: Fetching completed snarks from peer: $peer"
+        ~metadata:[ ("peer", Network_peer.Peer.to_yojson peer) ] ;
+      let completed_works =
+        Mina_networking.get_completed_checked_snarks network peer
+      in
+      let%bind completed_works = completed_works in
+      let completed_works =
+        match completed_works with
+        | Error e ->
+            [%log debug]
+              ~metadata:
+                [ ("peer", Network_peer.Peer.to_yojson peer)
+                ; ("error", Error_json.error_to_yojson e)
+                ]
+              "Failed to fetch completed snarks from peer: $error" ;
+            []
+        | Ok completed_works ->
+            completed_works
+      in
+
+      [%log debug]
+        ~metadata:
+          [ ("peer", Network_peer.Peer.to_yojson peer)
+          ; ("completed_works", `Int (List.length completed_works))
+          ]
+        "Fetched $completed_works completed snarks from peer: $peer" ;
+
+      (* keep reading from the frontier pipe until the new top block is greater than 2 more than the old top block.public_key
+         we do this to account for the block lag placed after the bootstrap process. The number 2 was chosen after manual testing.
+         1 did not yield successful verifcation of the polled snarks work.
+      *)
+      let rec wait_for_new_top_block old_top_block =
+        let new_block_length_received =
+          let open Mina_block in
+          Mina_numbers.Length.to_int @@ Mina_block.Header.blockchain_length
+          @@ Validation.header
+          @@ get_most_recent_valid_block ()
```

Comment:
This is new block received and not necessarily added to the frontier. You'll want to call `best_tip` function to get the latest block added to the frontier

Reply:
The alternative is storing the rpc response in memory for completed work and processing once the best tip is up to date. You cannot query the best tip and verify at the same time due to the sync delay. This was imo the most memory efficient solution. I'm made a decision to trade the time it takes in the verification loop for minimal memory usage.

Reply:
i will add this as well, it seems like a safer option. i chose this because I wanted this to be done as quickly as possible, but doing this with the transition frontier is the same amount of time since you need to have ledger information anyway.